### PR TITLE
Fix resizing hack; Enable AppVeyor (closes #586, closes #722, closes #764)

### LINKED ISSUE #586: `resizeWindow` functional tests fails locally if browser initially not maximized

### LINKED ISSUE #722: 2px resizing error in Firefox
Sometimes after resizing maximized Firefox window, the resulting client size area is 2px bigger than requested. Reproduced on my machine (just one time), @inikulin's, @kirovboris'es.

### LINKED ISSUE #764: Enable AppVeyor
For running `"local"` functional tests. It will help detect bugs like #722 and #586.

### PR DESCRIPTION
**Pull request by AndreyBelym**

### CODE REVIEW

**Comment by inikulin** [comment]
@AndreyBelym I believe we don't need to run AppVeyor tests for node 10. So, you can get rid of `path-is-absolute` dependency

**Comment by inikulin** [comment]
We don't have latest build bot with multiCI support deployed yet, right?

**Comment by AlexanderMoskovkin** [comment]
> We don't have latest build bot with multiCI support deployed yet, right?

Yes, it's [in progress](https://github.com/AlexanderMoskovkin/github-build-bot/pull/4)

File: test/functional/spawn.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+var execa     = require('execa');
+var Promise   = require('pinkie');
+var assign    = require('lodash').assign;
+
+
+module.exports = function (cmd, args, opts) {
+    opts = opts || {};
```
Comment (inikulin) [inline test/functional/spawn.js]: Maybe we just make a package for this already? =)

Comment (AndreyBelym) [inline test/functional/spawn.js]: Nevermind, actually there is no need for it here, I've figured how to use `execa`. BTW it's here just for debugging purposes, it will be removed just after I will be done with Firefox on AppVeyor.

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by inikulin** [comment]
I'm going to release RC today. Is there anything that blocking this from merging?

**Comment by AndreyBelym** [comment]
\cc @inikulin @AlexanderMoskovkin

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
QUnit tests `B253370 - Event handlers are called in the wrong order (click on option in select with size more than one)` and `B253370 - Event handlers are called in the wrong order (click on the same option in select with size more than one)` are failing in current `master` because of Chrome 53 update: https://github.com/DevExpress/testcafe/pull/786

**Comment by AlexanderMoskovkin** [comment]
@helen-dikareva please look at this asap

**Comment by helen-dikareva** [comment]
ok, I working on it

**Comment by inikulin** [comment]
What's the status of this one?

**Comment by AndreyBelym** [comment]
It's revealed that flaws in functional harness masked a test failure in MS Edge: https://github.com/DevExpress/testcafe/pull/793. We didn't caught it before, because we tested Edge and other MS browsers concurrently.

**Comment by inikulin** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by inikulin** [comment]
@AndreyBelym Why we have old bot reports now?

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
It's because Travis CI is disabled.

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
FPR

**Comment by inikulin** [comment]
OK, cool. But which browsers currently run in which test task? Can we rename the tasks to highlight it?
What's the difference between AppVeyor functional tests and Travis `test-functional-desktop` task?

**Comment by inikulin** [comment]
Also, I guess `test-functional-travis-ms-desktop` can be renamed to `test-functional-travis-ms-edge` now?

**Comment by inikulin** [comment]
Also, can we merge `edge` task with `test-functional-desktop`?

**Comment by AndreyBelym** [comment]
AppVeyor functional tests are local functional tests (with local browsers installed on AppVeyor's Win2012R2 VM), `test-functional-desktop` runs functional tests in common desktop browsers on Saucelabs.

**Comment by inikulin** [comment]
AppVeyor functional tests are local functional tests (with local browsers installed on AppVeyor's Win2012R2 VM), test-functional-desktop runs functional tests in common desktop browsers on Saucelabs.

So why we need both?

**Comment by AndreyBelym** [comment]
Because of macOS desktop browsers, and MS Edge. It is possible to use macOS VM on Travis to run local functional tests - https://github.com/DevExpress/testcafe/issues/791, but there is some screenshots issues that are preventing to do it right now. And MS Edge is going be likely not available until AppVeyor migrates to Win2016 (still TechPreview now).

**Comment by inikulin** [comment]
So, did I get it right:
on Travis we run browser on MacOS?

Can we have combined task `test-functional-travis-desktop-osx-and-ms-edge`?

**Comment by AndreyBelym** [comment]
FPR

File: src/browser/provider/built-in/base.js [hunk 2]
```diff
@@ -9,9 +9,15 @@ function getDataFromBrowser () {
...
 /*eslint-disable no-undef*/
 
-const INIT_SCRIPT = getDataFromBrowser.toString();
+const INIT_SCRIPT           = getDataFromBrowser.toString();
+const BROWSER_OPENING_DELAY = 3000;
+const RESIZE_DIFF_SIZE      = 100;
 
 
+function delay (ms) {
+    return new Promise(resolve => setTimeout(resolve, ms));
```
Comment (inikulin) [inline src/browser/provider/built-in/base.js]: I guess it can be moved to `utils`. Also, we need to import `Promise`.

File: src/browser/provider/built-in/base.js [hunk 3]
```diff
@@ -9,9 +9,15 @@ function getDataFromBrowser () {
 }
 /*eslint-disable no-undef*/
 
-const INIT_SCRIPT = getDataFromBrowser.toString();
+const INIT_SCRIPT           = getDataFromBrowser.toString();
+const BROWSER_OPENING_DELAY = 3000;
```
Comment (inikulin) [inline src/browser/provider/built-in/base.js]: Isn't it too long for bootstrapping?

**Comment by inikulin** [comment]
\r-

File: src/browser/provider/built-in/base.js (hunk 3 above)
Comment (AndreyBelym) [inline src/browser/provider/built-in/base.js]: Please note, that our client-side scripts are executed even before the browser window is opened, and it means that most part of this timeout the browser window will be just invisible. 

Also, window opening animations can take different amount of time depending on the machine performance. On my relatively fast machine it takes from 0.5 to 1 sec to play opening animation, and on a slower machine this time may be worse. 

I think, it's possible to decrease timeout to 2 secs, but lesser values may cause instability.

**Comment by AndreyBelym** [comment]
\cc @helen-dikareva

File: src/browser/provider/built-in/base.js [hunk 4]
```diff
@@ -22,23 +25,35 @@ export default class BrowserProviderBase {
...
 
         if (!await browserTools.isMaximized(title))
             return;
 
-        await browserTools.resize(title, width, height, width, height);
+        var testingWidth  = width - RESIZE_DIFF_SIZE;
+        var testingHeight = height - RESIZE_DIFF_SIZE;
 
-        var { width: newWidth, height: newHeight } = await this.runInitScript(browserId, INIT_SCRIPT);
+        this.resizeCorrections[browserId] = { width: 0, height: 0 };
```
Comment (helen-dikareva) [inline src/browser/provider/built-in/base.js]: We can do not init it and after rewrite, just init after first resize.

File: src/browser/provider/built-in/base.js [hunk 5]
```diff
@@ -22,23 +25,35 @@ export default class BrowserProviderBase {
...
 
-        await browserTools.maximize(title);
+        await browserTools.resize(title, width, height, testingWidth, testingHeight);
 
-        if (newWidth === width && newHeight === height)
-            return;
+        var newSize = await this.runInitScript(browserId, INIT_SCRIPT);
+
+        this.resizeCorrections[browserId].width += newSize.width - testingWidth;
```
Comment (helen-dikareva) [inline src/browser/provider/built-in/base.js]: We can extract these two lines in method

**Comment by AndreyBelym** [comment]
FPR

File: src/browser/provider/built-in/base.js [hunk 6]
```diff
@@ -1,16 +1,46 @@
...
+
+
+function sumSizes (sizeA, sizeB) {
+    return {
+        width:  sizeA.width + sizeB.width,
+        height: sizeA.height + sizeB.height
+    };
+}
 
+function substractSizes (sizeA, sizeB) {
```
Comment (helen-dikareva) [inline src/browser/provider/built-in/base.js]: subtract

**Comment by AndreyBelym** [comment]
\cc @AlexanderMoskovkin, please

**Comment by AlexanderMoskovkin** [comment]
lgtm
